jax/_src/lax/parallel.py

```diff
@@ -405,50 +405,77 @@ def pdot(x, y, axis_name, pos_contract=((), ()), pos_batch=((), ()),
-
-  lhs_reduce_axes = [lhs_subs.index(n) for n in lhs_uniques & subs_contract]
-  if lhs_reduce_axes:
-    x = lax._reduce_sum(x, lhs_reduce_axes)
-    for i in sorted(lhs_reduce_axes, reverse=True):
-      del lhs_subs[i]
-
-  rhs_reduce_axes = [rhs_subs.index(n) for n in rhs_uniques & subs_contract]
-  if rhs_reduce_axes:
-    y = lax._reduce_sum(y, rhs_reduce_axes)
-    for i in sorted(rhs_reduce_axes, reverse=True):
-      del rhs_subs[i]
-
-  pos_contract = unzip2((lhs_subs.index(n), rhs_subs.index(n))
-                        for n in subs_contract - (lhs_uniques | rhs_uniques))
-
-  # if a subscript apperas in both inputs _and_ the outputs, batch!
-  subs_batch = all_subs - subs_contract
-  if subs_batch & (lhs_uniques | rhs_uniques):
-    raise NotImplementedError
-
-  pos_batch = unzip2((lhs_subs.index(n), rhs_subs.index(n))
-                        for n in subs_batch)
-
-  return pdot(x, y, axis_name=named_contract,
-              pos_contract=pos_contract, pos_batch=pos_batch)
+
+  if len(operands) != len(all_in_named):
+    raise ValueError("Expecting the same number of argument specs in the "
+                     "subscript ({in_spec}) as the number of operands. But got "
+                     "{len(all_in_named)} argument specs for "
+                     "{len(operands)} operands")
+
+  if len(operands) > 2:
+    raise NotImplementedError("Only one or two operands are supported. "
+                              f"But got {len(operands)} operands")
+
+  # output subs and named axes must appear in at least one of the inputs.
+  if not set(out_named).issubset(set().union(*all_in_named)):
+    raise ValueError("Found named axes "
+                     f"{set(out_named) - set().union(*all_in_named)} "
+                     "appearing in the output spec but not in the input")
+  if not set(out_subs).issubset(set().union(*all_in_subs)):
+    raise ValueError("Found subscript(s) "
+                     f"{set(out_subs) - set().union(*all_in_subs)} "
+                     "appearing in the output spec but not in the input")
+
+  xs = list(operands)
+  for idx, (x, in_subs, in_named) in enumerate(zip(xs, all_in_subs, all_in_named)):
+    # if a subscript axis appears only in one input and not the output, reduce!
+    other_named = set().union(  # type: ignore
+        *[named for i, named in enumerate(all_in_named) if i != idx])
+    other_subs = set().union(  # type: ignore
+        *[subs for i, subs in enumerate(all_in_subs) if i != idx])
+
+    # reduce positional axes.
+    subs_reduce = list(set(in_subs) - {*out_subs, *other_subs})
+    subs_reduce_axes = [in_subs.index(n) for n in subs_reduce]
+    if subs_reduce_axes:
+      x = lax._reduce_sum(x, subs_reduce_axes)
+      for i in sorted(subs_reduce_axes, reverse=True):
+        del all_in_subs[idx][i]
+
+    # reduce named axes.
+    named_reduce_axes = list(set(in_named) - {*out_named, *other_named})
+    for ax in named_reduce_axes:
+      x = psum(x, ax)
```

Comment:
By the way you can do a `psum` with a tuple of both named and positional axes and it should work. That could simplify the code.

Reply:
Also, you delete the reduced axes from `all_in_subs` above, but you don't remove it from `all_in_named` here. Is that intentional?

Reply:
Yes, that was intentional. Removing from all_in_subs is necessary because we need to find the correct position for the subscript later on when contracting. But for all_in_named that's not necessary as we can just refer to the axis name in pdot.

It's another reason why named axes are cool (they're oblivious to other axes being removed)

Reply:
Re: doing psum with both named and positional axes -- thats a good point!

However, I thought `psum` can only take one named axis at a time? (While it can accept multiple positional axes in the same call) If that is true, we can't simplify much here.

Reply:
Sounds good. I'll try this out.

Reply:
Updated with a single psum and also now removing the reduced named axes. Thanks for the timely reviews!